# Preprocessing with scikit-learn - Cumulative Lab

## Introduction
In this cumulative lab, you'll practice applying various preprocessing techniques with scikit-learn (`sklearn`) to the Ames Housing dataset in order to prepare the data for predictive modeling. The main emphasis here is on preprocessing (not EDA or modeling theory), so we will skip over most of the visualization and metrics steps that you would take in an actual modeling process.

## Objectives

You will be able to:

* Practice identifying which preprocessing technique to use
* Practice filtering down to relevant columns
* Practice applying `sklearn.impute` to fill in missing values
* Practice applying `sklearn.preprocessing`:
  * `OrdinalEncoder` for converting binary categories to 0 and 1 within a single column
  * `OneHotEncoder` for creating multiple "dummy" columns to represent multiple categories

## Your Task: Prepare the Ames Housing Dataset for Modeling

![house in Ames](https://curriculum-content.s3.amazonaws.com/data-science/images/ames_house.jpg)

<span>Photo by <a href="https://unsplash.com/@kjkempt17?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Kyle Kempt</a> on <a href="https://unsplash.com/s/photos/ames?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">Unsplash</a></span>

### Requirements

#### 1. Drop Irrelevant Columns

For the purposes of this lab, we will only be using a subset of all of the features present in the Ames Housing dataset. In this step you will drop all irrelevant columns.

#### 2. Handle Missing Values

Often for reasons outside of a data scientist's control, datasets are missing some values. In this step you will assess the presence of NaN values in our subset of data, and use `MissingIndicator` and `SimpleImputer` from the `sklearn.impute` submodule to handle any missing values.

#### 3. Convert Categorical Features into Numbers

A built-in assumption of the scikit-learn library is that all data being fed into a machine learning model is already in a numeric format, otherwise you will get a `ValueError` when you try to fit a model. In this step you will use an `OrdinalEncoder` to replace data within individual non-numeric columns with 0s and 1s, and a `OneHotEncoder` to replace columns containing more than 2 categories with multiple "dummy" columns containing 0s and 1s.

At this point, a scikit-learn model should be able to run without errors!

#### 4. Preprocess Test Data

Apply Steps 1-3 to the test data in order to perform a final model evaluation.

## Lab Setup

### Getting the Data

In the cell below, we import the `pandas` library, open the CSV containing the Ames Housing data as a pandas `DataFrame`, and inspect its contents.

In [49]:
# Run this cell without changes
import pandas as pd
df = pd.read_csv("data/ames.csv")
df

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [50]:
# Run this cell without changes
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

The prediction target for this analysis is the sale price of the home, so we separate the data into `X` and `y` accordingly:

In [51]:
# Run this cell without changes
y = df["SalePrice"]
X = df.drop("SalePrice", axis=1)

Next, we separate the data into a train set and a test set prior to performing any preprocessing steps:

In [52]:
# Run this cell without changes
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

(If you are working through this lab and you just want to start over with the original value for `X_train`, re-run the cell above.)

In [53]:
# Run this cell without changes
print(f"X_train is a DataFrame with {X_train.shape[0]} rows and {X_train.shape[1]} columns")
print(f"y_train is a Series with {y_train.shape[0]} values")

# We always should have the same number of rows in X as values in y
assert X_train.shape[0] == y_train.shape[0]

X_train is a DataFrame with 1095 rows and 80 columns
y_train is a Series with 1095 values


#### Fitting a Model

For this lab we will be using a `LinearRegression` model from scikit-learn ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)).

Right now, we have not done any preprocessing, so we expect that trying to fit a model will fail:

In [54]:
# Run this cell without changes
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

ValueError: could not convert string to float: 'RL'

As you can see, we got `ValueError: could not convert string to float: 'RL'`.

In order to fit a scikit-learn model, all values must be numeric, and the third column of our full dataset (`MSZoning`) contains values like `'RL'` and `'RH'`, which are strings. So this error was expected, but after some preprocessing, this model will work!

## 1. Drop Irrelevant Columns

For the purpose of this analysis, we'll only use the following columns, described by `relevant_columns`. You can find the full description of their values in the file `data/data_description.txt` included in this repository.

In the cell below, reassign `X_train` so that it only contains the columns in `relevant_columns`.

**Hint:** Even though we describe this as "dropping" irrelevant columns, it's easier if you invert the logic, so that we are only keeping relevant columns, rather than using the `.drop()` method. It is possible to use the `.drop()` method if you really want to, but first you would need to create a list of the column names that you don't want to keep.

In [ ]:
# Replace None with appropriate code

# Declare relevant columns
relevant_columns = [
    'LotFrontage',  # Linear feet of street connected to property
    'LotArea',      # Lot size in square feet
    'Street',       # Type of road access to property
    'OverallQual',  # Rates the overall material and finish of the house
    'OverallCond',  # Rates the overall condition of the house
    'YearBuilt',    # Original construction date
    'YearRemodAdd', # Remodel date (same as construction date if no remodeling or additions)
    'GrLivArea',    # Above grade (ground) living area square feet
    'FullBath',     # Full bathrooms above grade
    'BedroomAbvGr', # Bedrooms above grade (does NOT include basement bedrooms)
    'TotRmsAbvGrd', # Total rooms above grade (does not include bathrooms)
    'Fireplaces',   # Number of fireplaces
    'FireplaceQu',  # Fireplace quality
    'MoSold',       # Month Sold (MM)
    'YrSold'        # Year Sold (YYYY)
]

# Reassign X_train so that it only contains relevant columns
X_train = X_train.loc[:, relevant_columns]

# Visually inspect X_train
X_train

Check that the new shape is correct:

In [ ]:
# Run this cell without changes

# X_train should have the same number of rows as before
assert X_train.shape[0] == 1095

# Now X_train should only have as many columns as relevant_columns
assert X_train.shape[1] == len(relevant_columns)

## 2. Handle Missing Values

In the cell below, we check to see if there are any NaNs in the selected subset of data:

In [ ]:
# Run this cell without changes
X_train.isna().sum()

Ok, it looks like we have some NaNs in `LotFrontage` and `FireplaceQu`.

Before we proceed to fill in those values, we need to ask: **do these NaNs actually represent** ***missing*** **values, or is there some real value/category being represented by NaN?**

### Fireplace Quality

To start with, let's look at `FireplaceQu`, which means "Fireplace Quality". Why might we have NaN fireplace quality?

Well, some properties don't have fireplaces!

Let's confirm this guess with a little more analysis.

First, we know that there are 512 records with NaN fireplace quality. How many records are there with zero fireplaces?

In [ ]:
# Run this cell without changes
X_train[X_train["Fireplaces"] == 0]

Ok, that's 512 rows, same as the number of NaN `FireplaceQu` records. To double-check, let's query for that combination of factors (zero fireplaces and `FireplaceQu` is NaN):

In [ ]:
# Run this cell without changes
X_train[
    (X_train["Fireplaces"] == 0) &
    (X_train["FireplaceQu"].isna())
]

Looks good, still 512 records. So, NaN fireplace quality is not actually information that is missing from our dataset, it is a genuine category which means "fireplace quality is not applicable". This interpretation aligns with what we see in `data/data_description.txt`:

```
...
FireplaceQu: Fireplace quality

       Ex	Excellent - Exceptional Masonry Fireplace
       Gd	Good - Masonry Fireplace in main level
       TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement
       Fa	Fair - Prefabricated Fireplace in basement
       Po	Poor - Ben Franklin Stove
       NA	No Fireplace
...
```

So, let's replace those NaNs with the string "N/A" to indicate that this is a real category, not missing data:

In [ ]:
# Run this cell without changes
X_train["FireplaceQu"] = X_train["FireplaceQu"].fillna("N/A")
X_train["FireplaceQu"].value_counts()

Eventually we will still need to perform some preprocessing to prepare the `FireplaceQu` column for modeling (because models require numeric inputs rather than inputs of type `object`), but we don't need to worry about filling in missing values.

### Lot Frontage

Now let's look at `LotFrontage` — it's possible that NaN is also a genuine category here, and it's possible that it's just missing data instead. Let's apply some domain understanding to understand whether it's possible that lot frontage can be N/A just like fireplace quality can be N/A.

Lot frontage is defined as the "Linear feet of street connected to property", i.e. how much of the property runs directly along a road. The amount of frontage required for a property depends on its zoning. Let's look at the zoning of all records with NaN for `LotFrontage`:

In [55]:
# Run this cell without changes
df[df["LotFrontage"].isna()]["MSZoning"].value_counts()

MSZoning
RL    229
RM     19
FV      8
RH      3
Name: count, dtype: int64

So, we have RL (residential low density), RM (residential medium density), FV (floating village residential), and RH (residential high density). Looking at the building codes from the City of Ames, it appears that all of these zones require at least 24 feet of frontage.

Nevertheless, we can't assume that all properties have frontage just because the zoning regulations require it. Maybe these properties predate the regulations, or they received some kind of variance permitting them to get around the requirement. **It's still not as clear here as it was with the fireplaces whether this is a genuine "not applicable" kind of NaN or a "missing information" kind of NaN.**

In a case like this, we can take a double approach:

1. Make a new column in the dataset that simply represents whether `LotFrontage` was originally NaN
2. Fill in the NaN values of `LotFrontage` with median frontage in preparation for modeling

### Missing Indicator for `LotFrontage`

First, we import `sklearn.impute.MissingIndicator` ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.impute.MissingIndicator.html)). The goal of using a `MissingIndicator` is creating a new column to represent which values were NaN (or some other "missing" value) in the original dataset, in case NaN ends up being a meaningful indicator rather than a random missing bit of data.

A `MissingIndicator` is a scikit-learn transformer, meaning that we will use the standard steps for any scikit-learn transformer:

1. Identify data to be transformed (typically not every column is passed to every transformer)
2. Instantiate the transformer object
3. Fit the transformer object (on training data only)
4. Transform data using the transformer object
5. Add the transformed data to the other data that was not transformed

In [56]:
# Replace None with appropriate code
from sklearn.impute import MissingIndicator

# (1) Identify data to be transformed
# We only want missing indicators for LotFrontage
frontage_train = X_train[["LotFrontage"]]

# (2) Instantiate the transformer object
missing_indicator = MissingIndicator()

# (3) Fit the transformer object on frontage_train
missing_indicator.fit(frontage_train)

# (4) Transform frontage_train and assign the result
# to frontage_missing_train
frontage_missing_train = missing_indicator.transform(frontage_train)

# Visually inspect frontage_missing_train
frontage_missing_train

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

The result of transforming `frontage_train` should be an array of arrays, each containing `True` or `False`. Make sure the `assert`s pass before moving on to the next step.

In [57]:
# Run this cell without changes
import numpy as np

# frontage_missing_train should be a NumPy array
assert type(frontage_missing_train) == np.ndarray

# We should have the same number of rows as the full X_train
assert frontage_missing_train.shape[0] == X_train.shape[0]

# But we should only have 1 column
assert frontage_missing_train.shape[1] == 1

Now let's add this new information as a new column of `X_train`:

In [58]:
# Run this cell without changes

# (5) add the transformed data to the other data
X_train["LotFrontage_Missing"] = frontage_missing_train
X_train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1023  1024         120       RL         43.0     3182   Pave   NaN      Reg   
810    811          20       RL         78.0    10140   Pave   NaN      Reg   
1384  1385          50       RL         60.0     9060   Pave   NaN      Reg   
626    627          20       RL          NaN    12342   Pave   NaN      IR1   
813    814          20       RL         75.0     9750   Pave   NaN      Reg   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1095  1096          20       RL         78.0     9317   Pave   NaN      IR1   
1130  1131          50       RL         65.0     7804   Pave   NaN      Reg   
1294  1295          20       RL         60.0     8172   Pave   NaN      Reg   
860    861          50       RL         55.0     7642   Pave   NaN      Reg   
1126  1127         120       RL         53.0     3684   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
1023         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
810          Lvl    AllPub  ...      648     Fa  GdPrv         NaN       0   
1384         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
626          Lvl    AllPub  ...        0    NaN   GdWo        Shed     600   
813          Lvl    AllPub  ...        0    NaN    NaN        Shed     500   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1095         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1130         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1294         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
860          Lvl    AllPub  ...        0    NaN  GdPrv         NaN       0   
1126         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  LotFrontage_Missing  
1023      5   2008        WD         Normal                False  
810       1   2006        WD         Normal                False  
1384     10   2009        WD         Normal                False  
626       8   2007        WD         Normal                 True  
813       4   2007       COD         Normal                False  
...     ...    ...       ...            ...                  ...  
1095      3   2007        WD         Normal                False  
1130     12   2009        WD         Normal                False  
1294      4   2006        WD         Normal                False  
860       6   2007        WD         Normal                False  
1126      6   2009        WD         Normal                False  

[1095 rows x 81 columns]

In [59]:
# Run this cell without changes

# Now we should have 1 extra column compared to
# our original subset
assert X_train.shape[1] == len(relevant_columns) + 1

AssertionError: 

### Imputing Missing Values for LotFrontage

Now that we have noted where missing values were originally present, let's use a `SimpleImputer` ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html)) to fill in those NaNs in the `LotFrontage` column.

The process is very similar to the `MissingIndicator` process, except that we want to replace the original `LotFrontage` column with the transformed version instead of just adding a new column on.

In the cell below, create and use a `SimpleImputer` with `strategy="median"` to transform the value of `frontage_train` (declared above).

In [ ]:
# Replace None with appropriate code

from sklearn.impute import SimpleImputer

# (1) frontage_train was created previously, so we don't
# need to extract the relevant data again

# (2) Instantiate a SimpleImputer with strategy="median"
imputer = SimpleImputer(strategy="median")

# (3) Fit the imputer on frontage_train
imputer.fit(frontage_train)

# (4) Transform frontage_train using the imputer and
# assign the result to frontage_imputed_train
frontage_imputed_train = imputer.transform(frontage_train)

# Visually inspect frontage_imputed_train
frontage_imputed_train

Now we can replace the original value of `LotFrontage` in `X_train` with the new value:

In [ ]:
# Run this cell without changes

# (5) Replace value of LotFrontage
X_train["LotFrontage"] = frontage_imputed_train

# Visually inspect X_train
X_train

Now the shape of `X_train` should still be the same as before:

In [ ]:
# Run this cell without changes
assert X_train.shape == (1095, 16)

And now our `X_train` no longer contains any NaN values:

In [60]:
# Run this cell without changes
X_train.isna().sum()

Id                       0
MSSubClass               0
MSZoning                 0
LotFrontage            200
LotArea                  0
                      ... 
MoSold                   0
YrSold                   0
SaleType                 0
SaleCondition            0
LotFrontage_Missing      0
Length: 81, dtype: int64

Great! Now we have completed Step 2.

## 3. Convert Categorical Features into Numbers

Despite dropping irrelevant columns and filling in those NaN values, if we feed the current `X_train` into our scikit-learn `LinearRegression` model, it will crash:

In [61]:
# Run this cell without changes
model.fit(X_train, y_train)

ValueError: could not convert string to float: 'RL'

Now the first column to cause a problem is `Street`, which is documented like this:

```
...
Street: Type of road access to property

       Grvl	Gravel	
       Pave	Paved
...
```

Let's look at the full `X_train`:

In [ ]:
# Run this cell without changes
X_train.info()

So, our model is crashing because some of the columns are non-numeric.

Anything that is already `float64` or `int64` will work with our model, but these features need to be converted:

* `Street` (currently type `object`)
* `FireplaceQu` (currently type `object`)
* `LotFrontage_Missing` (currently type `bool`)

There are two main approaches to converting these values, depending on whether there are 2 values (meaning the categorical variable can be converted into a single binary number) or more than 2 values (meaning we need to create extra columns to represent all categories).

(If there is only 1 value, this is not a useful feature for the purposes of predictive analysis because every single row contains the same information.)

In the cell below, we inspect the value counts of the specified features:

In [ ]:
# Run this cell without changes

print(X_train["Street"].value_counts())
print()
print(X_train["FireplaceQu"].value_counts())
print()
print(X_train["LotFrontage_Missing"].value_counts())

So, it looks like `Street` and `LotFrontage_Missing` have only 2 categories and can be converted into binary in place, whereas `FireplaceQu` has 6 categories and will need to be expanded into multiple columns.

### Binary Categories

For binary categories, we will use an `OrdinalEncoder` ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html)) to convert the categories of `Street` and `LotFrontage_Missing` into binary values (0s and 1s).

Just like in Step 2 when we used the `MissingIndicator` and `SimpleImputer`, we will follow these steps:

1. Identify data to be transformed
2. Instantiate the transformer object
3. Fit the transformer object (on training data only)
4. Transform data using the transformer object
5. Add the transformed data to the other data that was not transformed

Let's start with transforming `Street`:

In [ ]:
# Replace None with appropriate code

# (0) import OrdinalEncoder from sklearn.preprocessing
from sklearn.preprocessing import OrdinalEncoder

# (1) Create a variable street_train that contains the
# relevant column from X_train
# (Use double brackets [[]] to get the appropriate shape)
street_train = X_train[["Street"]]

# (2) Instantiate an OrdinalEncoder
encoder_street = OrdinalEncoder()

# (3) Fit the encoder on street_train
encoder_street.fit(street_train)

# Inspect the categories of the fitted encoder
encoder_street.categories_[0]

The `.categories_` attribute of `OrdinalEncoder` is only present once the `.fit` method has been called. (The trailing `_` indicates this convention.)

What this tells us is that when `encoder_street` is used to transform the street data into 1s and 0s, `0` will mean `'Grvl'` (gravel) in the original data, and `1` will mean `'Pave'` (paved) in the original data.

The eventual scikit-learn model only cares about the 1s and 0s, but this information can be useful for us to understand what our code is doing and help us debug when things go wrong.

Now let's transform `street_train` with the fitted encoder:

In [ ]:
# Replace None with appropriate code

# (4) Transform street_train using the encoder and
# assign the result to street_encoded_train
street_encoded_train = encoder_street.transform(street_train)

# Flatten for appropriate shape
street_encoded_train = street_encoded_train.flatten()

# Visually inspect street_encoded_train
street_encoded_train

All of the values we see appear to be `1` right now, but that makes sense since there were only 4 properties with gravel (`0`) streets in `X_train`.

Now let's replace the original `Street` column with the encoded version:

In [ ]:
# Replace None with appropriate code

# (5) Replace value of Street
X_train["Street"] = street_encoded_train

# Visually inspect X_train
X_train

In [ ]:
# Run this cell without changes
X_train.info()

Perfect! Now `Street` should by type `int64` instead of `object`.

Now, repeat the same process with `LotFrontage_Missing`:

In [ ]:
# Replace None with appropriate code

# (1) We already have a variable frontage_missing_train
# from earlier, no additional step needed

# (2) Instantiate an OrdinalEncoder for missing frontage
encoder_frontage_missing = OrdinalEncoder()

# (3) Fit the encoder on frontage_missing_train
encoder_frontage_missing.fit(frontage_missing_train)

# Inspect the categories of the fitted encoder
encoder_frontage_missing.categories_[0]

In [ ]:
# Replace None with appropriate code

# (4) Transform frontage_missing_train using the encoder and
# assign the result to frontage_missing_encoded_train
frontage_missing_encoded_train = encoder_frontage_missing.transform(frontage_missing_train)

# Flatten for appropriate shape
frontage_missing_encoded_train = frontage_missing_encoded_train.flatten()

# Visually inspect frontage_missing_encoded_train
frontage_missing_encoded_train

In [ ]:
# Replace None with appropriate code

# (5) Replace value of LotFrontage_Missing
X_train["LotFrontage_Missing"] = frontage_missing_encoded_train

# Visually inspect X_train
X_train

In [ ]:
# Run this cell without changes
X_train.info()

Great, now we only have 1 column remaining that isn't type `float64` or `int64`!

#### Note on Preprocessing Boolean Values
For binary values like `LotFrontage_Missing`, you might see a few different approaches to preprocessing. Python treats `True` and `1` as equal:

In [ ]:
# Run this cell without changes
print(True == 1)
print(False == 0)

This means that if your model is purely using Python, you actually might just be able to leave columns as type `bool` without any issues. You will likely see examples that do this. However if your model relies on C or Java "under the hood", this might cause problems.

There is also a technique using `pandas` rather than scikit-learn for this particular conversion of boolean values to integers:

In [ ]:
# Run this cell without changes
df_example = pd.DataFrame(frontage_missing_train, columns=["LotFrontage_Missing"])
df_example

In [ ]:
# Run this cell without changes
df_example["LotFrontage_Missing"] = df_example["LotFrontage_Missing"].astype(int)
df_example

This code is casting every value in the `LotFrontage_Missing` column to an integer, achieving the same result as the `OrdinalEncoder` example with less code.

The downside of using this approach is that it doesn't fit into a scikit-learn pipeline as neatly because it is using `pandas` to do the transformation instead of scikit-learn.

In the future, you will need to make your own determination of which strategy to use!

### Multiple Categories

Unlike `Street` and `LotFrontage_Missing`, `FireplaceQu` has more than two categories. Therefore the process for encoding it numerically is a bit more complicated, because we will need to create multiple "dummy" columns that are each representing one category.

To do this, we can use a `OneHotEncoder` from `sklearn.preprocessing` ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html)).

The first several steps are very similar to all of the other transformers we've used so far, although the process of combining the data with the original data differs.

In the cells below, complete steps `(0)`-`(4)` of preprocessing the `FireplaceQu` column using a `OneHotEncoder`:

In [ ]:
# Replace None with appropriate code

# (0) import OneHotEncoder from sklearn.preprocessing
from sklearn.preprocessing import OneHotEncoder

# (1) Create a variable fireplace_qu_train
# extracted from X_train
# (double brackets due to shape expected by OHE)
fireplace_qu_train = X_train[["FireplaceQu"]]

# (2) Instantiate a OneHotEncoder with categories="auto",
# sparse=False, and handle_unknown="ignore"
ohe = OneHotEncoder(categories="auto", sparse=False, handle_unknown="ignore")

# (3) Fit the encoder on fireplace_qu_train
ohe.fit(fireplace_qu_train)

# Inspect the categories of the fitted encoder
ohe.categories_

In [ ]:
# Replace None with appropriate code

# (4) Transform fireplace_qu_train using the encoder and
# assign the result to fireplace_qu_encoded_train
fireplace_qu_encoded_train = ohe.transform(fireplace_qu_train)

# Visually inspect fireplace_qu_encoded_train
fireplace_qu_encoded_train

Notice that this time, unlike with `MissingIndicator`, `SimpleImputer`, or `OrdinalEncoder`, we have created multiple columns of data out of a single column. The code below converts this unlabeled NumPy array into a readable pandas dataframe in preparation for merging it back with the rest of `X_train`:

In [62]:
# Run this cell without changes

# (5a) Make the transformed data into a dataframe
fireplace_qu_encoded_train = pd.DataFrame(
    # Pass in NumPy array
    fireplace_qu_encoded_train,
    # Set the column names to the categories found by OHE
    columns=ohe.categories_[0],
    # Set the index to match X_train's index
    index=X_train.index
)

# Visually inspect new dataframe
fireplace_qu_encoded_train

Ex   Fa   Gd  N/A   Po   TA
1023  0.0  0.0  1.0  0.0  0.0  0.0
810   0.0  1.0  0.0  0.0  0.0  0.0
1384  0.0  0.0  0.0  1.0  0.0  0.0
626   0.0  0.0  0.0  0.0  0.0  1.0
813   0.0  0.0  0.0  1.0  0.0  0.0
...   ...  ...  ...  ...  ...  ...
1095  0.0  0.0  1.0  0.0  0.0  0.0
1130  0.0  0.0  0.0  0.0  0.0  1.0
1294  0.0  0.0  0.0  1.0  0.0  0.0
860   0.0  0.0  1.0  0.0  0.0  0.0
1126  0.0  0.0  0.0  0.0  0.0  1.0

[1095 rows x 6 columns]

A couple notes on the code above:

* The main goal of converting this into a dataframe (rather than converting `X_train` into a NumPy array, which would also allow them to be combined) is **readability** — to help you and others understand what your code is doing, and to help you debug. Eventually when you write this code as a pipeline, it will be NumPy arrays "under the hood".
* We are using just the **raw categories** from `FireplaceQu` as our new dataframe columns, but you'll also see examples where a lambda function or list comprehension is used to create column names indicating the original column name, e.g. `FireplaceQu_Ex`, `FireplaceQu_Fa` rather than just `Ex`, `Fa`. This is a design decision based on readability — the scikit-learn model will work the same either way.
* It is very important that **the index of the new dataframe matches the index of the main `X_train` dataframe**. Because we used `train_test_split`, the index of `X_train` is shuffled, so it goes `1023`, `810`, `1384` etc. instead of `0`, `1`, `2`, etc. If you don't specify an index for the new dataframe, it will assign the first record to the index `0` rather than `1023`. If you are ever merging encoded data like this and a bunch of NaNs start appearing, make sure that the indexes are lined up correctly! You also may see examples where the index of `X_train` has been reset, rather than specifying the index of the new dataframe — either way works.

Next, we want to drop the original `FireplaceQu` column containing the categorical data:

(For previous transformations we didn't need to drop anything because we were replacing 1 column with 1 new column in place, but one-hot encoding works differently.)

In [63]:
# Run this cell without changes

# (5b) Drop original FireplaceQu column
X_train.drop("FireplaceQu", axis=1, inplace=True)

# Visually inspect X_train
X_train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1023  1024         120       RL         43.0     3182   Pave   NaN      Reg   
810    811          20       RL         78.0    10140   Pave   NaN      Reg   
1384  1385          50       RL         60.0     9060   Pave   NaN      Reg   
626    627          20       RL          NaN    12342   Pave   NaN      IR1   
813    814          20       RL         75.0     9750   Pave   NaN      Reg   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1095  1096          20       RL         78.0     9317   Pave   NaN      IR1   
1130  1131          50       RL         65.0     7804   Pave   NaN      Reg   
1294  1295          20       RL         60.0     8172   Pave   NaN      Reg   
860    861          50       RL         55.0     7642   Pave   NaN      Reg   
1126  1127         120       RL         53.0     3684   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
1023         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
810          Lvl    AllPub  ...      648     Fa  GdPrv         NaN       0   
1384         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
626          Lvl    AllPub  ...        0    NaN   GdWo        Shed     600   
813          Lvl    AllPub  ...        0    NaN    NaN        Shed     500   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1095         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1130         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1294         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
860          Lvl    AllPub  ...        0    NaN  GdPrv         NaN       0   
1126         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  LotFrontage_Missing  
1023      5   2008        WD         Normal                False  
810       1   2006        WD         Normal                False  
1384     10   2009        WD         Normal                False  
626       8   2007        WD         Normal                 True  
813       4   2007       COD         Normal                False  
...     ...    ...       ...            ...                  ...  
1095      3   2007        WD         Normal                False  
1130     12   2009        WD         Normal                False  
1294      4   2006        WD         Normal                False  
860       6   2007        WD         Normal                False  
1126      6   2009        WD         Normal                False  

[1095 rows x 80 columns]

Finally, we want to concatenate the new dataframe together with the original `X_train`:

In [64]:
# Run this cell without changes

# (5c) Concatenate the new dataframe with current X_train
X_train = pd.concat([X_train, fireplace_qu_encoded_train], axis=1)

# Visually inspect X_train
X_train

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1023  1024         120       RL         43.0     3182   Pave   NaN      Reg   
810    811          20       RL         78.0    10140   Pave   NaN      Reg   
1384  1385          50       RL         60.0     9060   Pave   NaN      Reg   
626    627          20       RL          NaN    12342   Pave   NaN      IR1   
813    814          20       RL         75.0     9750   Pave   NaN      Reg   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1095  1096          20       RL         78.0     9317   Pave   NaN      IR1   
1130  1131          50       RL         65.0     7804   Pave   NaN      Reg   
1294  1295          20       RL         60.0     8172   Pave   NaN      Reg   
860    861          50       RL         55.0     7642   Pave   NaN      Reg   
1126  1127         120       RL         53.0     3684   Pave   NaN      Reg   

     LandContour Utilities  ... YrSold SaleType SaleCondition  \
1023         Lvl    AllPub  ...   2008       WD        Normal   
810          Lvl    AllPub  ...   2006       WD        Normal   
1384         Lvl    AllPub  ...   2009       WD        Normal   
626          Lvl    AllPub  ...   2007       WD        Normal   
813          Lvl    AllPub  ...   2007      COD        Normal   
...          ...       ...  ...    ...      ...           ...   
1095         Lvl    AllPub  ...   2007       WD        Normal   
1130         Lvl    AllPub  ...   2009       WD        Normal   
1294         Lvl    AllPub  ...   2006       WD        Normal   
860          Lvl    AllPub  ...   2007       WD        Normal   
1126         Lvl    AllPub  ...   2009       WD        Normal   

     LotFrontage_Missing   Ex   Fa   Gd  N/A   Po   TA  
1023               False  0.0  0.0  1.0  0.0  0.0  0.0  
810                False  0.0  1.0  0.0  0.0  0.0  0.0  
1384               False  0.0  0.0  0.0  1.0  0.0  0.0  
626                 True  0.0  0.0  0.0  0.0  0.0  1.0  
813                False  0.0  0.0  0.0  1.0  0.0  0.0  
...                  ...  ...  ...  ...  ...  ...  ...  
1095               False  0.0  0.0  1.0  0.0  0.0  0.0  
1130               False  0.0  0.0  0.0  0.0  0.0  1.0  
1294               False  0.0  0.0  0.0  1.0  0.0  0.0  
860                False  0.0  0.0  1.0  0.0  0.0  0.0  
1126               False  0.0  0.0  0.0  0.0  0.0  1.0  

[1095 rows x 86 columns]

In [65]:
# Run this cell without changes
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1095 entries, 1023 to 1126
Data columns (total 86 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   1095 non-null   int64  
 1   MSSubClass           1095 non-null   int64  
 2   MSZoning             1095 non-null   object 
 3   LotFrontage          895 non-null    float64
 4   LotArea              1095 non-null   int64  
 5   Street               1095 non-null   object 
 6   Alley                70 non-null     object 
 7   LotShape             1095 non-null   object 
 8   LandContour          1095 non-null   object 
 9   Utilities            1095 non-null   object 
 10  LotConfig            1095 non-null   object 
 11  LandSlope            1095 non-null   object 
 12  Neighborhood         1095 non-null   object 
 13  Condition1           1095 non-null   object 
 14  Condition2           1095 non-null   object 
 15  BldgType             1095 non-null   obj

Ok, everything is numeric now! We have completed the minimum necessary preprocessing to use these features in a scikit-learn model!

In [66]:
# Run this cell without changes
model.fit(X_train, y_train)

ValueError: could not convert string to float: 'RL'

Great, no error this time.

Let's use cross validation to take a look at the model's performance:

In [67]:
# Run this cell without changes
from sklearn.model_selection import cross_val_score

cross_val_score(model, X_train, y_train, cv=3)

ValueError: 
All the 3 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
3 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 732, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\base.py", line 1151, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\linear_model\_base.py", line 678, in fit
    X, y = self._validate_data(
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\base.py", line 621, in _validate_data
    X, y = check_X_y(X, y, **check_params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\utils\validation.py", line 1147, in check_X_y
    X = check_array(
        ^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\utils\validation.py", line 838, in check_array
    array = array.astype(new_dtype)
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\generic.py", line 6324, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\internals\managers.py", line 451, in astype
    return self.apply(
           ^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\internals\managers.py", line 352, in apply
    applied = getattr(b, f)(**kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\internals\blocks.py", line 511, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\dtypes\astype.py", line 242, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\dtypes\astype.py", line 187, in astype_array
    values = _astype_nansafe(values, dtype, copy=copy)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\sparr\anaconda3\Lib\site-packages\pandas\core\dtypes\astype.py", line 138, in _astype_nansafe
    return arr.astype(dtype, copy=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: could not convert string to float: 'RL'


Not terrible, we are explaining between 66% and 80% of the variance in the target with our current feature set. Let's say that this is our final model and move on to preparing the test data.

## 4. Preprocess Test Data

> Apply Steps 1-3 to the test data in order to perform a final model evaluation.

This part is done for you, and it should work automatically, assuming you didn't change the names of any of the transformer objects. Note that we are intentionally **not instantiating or fitting the transformers** here, because you always want to fit transformers on the training data only.

*Step 1: Drop Irrelevant Columns*

In [68]:
# Run this cell without changes
X_test = X_test.loc[:, relevant_columns]

*Step 2: Handle Missing Values*

In [69]:
# Run this cell without changes

# Replace FireplaceQu NaNs with "N/A"s
X_test["FireplaceQu"] = X_test["FireplaceQu"].fillna("N/A")

# Add missing indicator for lot frontage
frontage_test = X_test[["LotFrontage"]]
frontage_missing_test = missing_indicator.transform(frontage_test)
X_test["LotFrontage_Missing"] = frontage_missing_test

# Impute missing lot frontage values
frontage_imputed_test = imputer.transform(frontage_test)
X_test["LotFrontage"] = frontage_imputed_test

# Check that there are no more missing values
X_test.isna().sum()

LotFrontage            0
LotArea                0
Street                 0
OverallQual            0
OverallCond            0
YearBuilt              0
YearRemodAdd           0
GrLivArea              0
FullBath               0
BedroomAbvGr           0
TotRmsAbvGrd           0
Fireplaces             0
FireplaceQu            0
MoSold                 0
YrSold                 0
LotFrontage_Missing    0
dtype: int64

*Step 3: Convert Categorical Features into Numbers*

In [70]:
# Run this cell without changes

# Encode street type
street_test = X_test[["Street"]]
street_encoded_test = encoder_street.transform(street_test).flatten()
X_test["Street"] = street_encoded_test

# Encode frontage missing
frontage_missing_test = X_test[["LotFrontage_Missing"]]
frontage_missing_encoded_test = encoder_frontage_missing.transform(frontage_missing_test).flatten()
X_test["LotFrontage_Missing"] = frontage_missing_encoded_test

# One-hot encode fireplace quality
fireplace_qu_test = X_test[["FireplaceQu"]]
fireplace_qu_encoded_test = ohe.transform(fireplace_qu_test)
fireplace_qu_encoded_test = pd.DataFrame(
    fireplace_qu_encoded_test,
    columns=ohe.categories_[0],
    index=X_test.index
)
X_test.drop("FireplaceQu", axis=1, inplace=True)
X_test = pd.concat([X_test, fireplace_qu_encoded_test], axis=1)

# Visually inspect X_test
X_test

C:\Users\sparr\anaconda3\Lib\site-packages\sklearn\base.py:457: UserWarning: X has feature names, but OrdinalEncoder was fitted without feature names
  warnings.warn(


LotFrontage  LotArea  Street  OverallQual  OverallCond  YearBuilt  \
892          70.0     8414     1.0            6            8       1963   
1105         98.0    12256     1.0            8            5       1994   
413          56.0     8960     1.0            5            6       1927   
522          50.0     5000     1.0            6            7       1947   
1036         89.0    12898     1.0            9            5       2007   
...           ...      ...     ...          ...          ...        ...   
988          70.0    12046     1.0            6            6       1976   
243          75.0    10762     1.0            6            6       1980   
1342         70.0     9375     1.0            8            5       2002   
1057         70.0    29959     1.0            7            6       1994   
1418         71.0     9204     1.0            5            5       1963   

      YearRemodAdd  GrLivArea  FullBath  BedroomAbvGr  ...  Fireplaces  \
892           2003       1068         1             3  ...           0   
1105          1995       2622         2             3  ...           2   
413           1950       1028         1             2  ...           1   
522           1950       1664         2             3  ...           2   
1036          2008       1620         2             2  ...           1   
...            ...        ...       ...           ...  ...         ...   
988           1976       2030         2             4  ...           1   
243           1980       1217         1             3  ...           1   
1342          2002       2169         2             3  ...           1   
1057          1994       1850         2             3  ...           1   
1418          1963       1144         1             3  ...           0   

      MoSold  YrSold  LotFrontage_Missing   Ex   Fa   Gd  N/A   Po   TA  
892        2    2006                  0.0  0.0  0.0  0.0  1.0  0.0  0.0  
1105       4    2010                  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
413        3    2010                  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
522       10    2006                  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
1036       9    2009                  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
...      ...     ...                  ...  ...  ...  ...  ...  ...  ...  
988        6    2007                  1.0  0.0  0.0  0.0  0.0  0.0  1.0  
243        4    2009                  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
1342       8    2007                  1.0  0.0  0.0  1.0  0.0  0.0  0.0  
1057       1    2009                  1.0  0.0  0.0  1.0  0.0  0.0  0.0  
1418       8    2008                  0.0  0.0  0.0  0.0  1.0  0.0  0.0  

[365 rows x 21 columns]

Fit the model on the full training set, evaluate on test set:

In [71]:
# Run this cell without changes
model.fit(X_train, y_train)
model.score(X_test, y_test)

ValueError: could not convert string to float: 'RL'

Great, that worked! Now we have completed the full process of preprocessing the Ames Housing data in preparation for machine learning!

## Summary

In this cumulative lab, you used various techniques to prepare the Ames Housing data for modeling. You filtered down the full dataset to only relevant columns, filled in missing values, and converted categorical data into numeric data. Each time, you practiced the scikit-learn transformer workflow by instantiating the transformer, fitting on the relevant training data, transforming the training data, and transforming the test data at the end (without re-instantiating or re-fitting the transformer object).